src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/LanguageServer.cs

```diff
@@ -415,6 +422,152 @@ private void HandleCodeActionRequest(string id, string paramsJson)
             _sendToClient(JsonRpcHelper.CreateSuccessResult(id, codeActions));
         }
 
+        /// <summary>
+        /// Handles requests to compute semantic requests for the full document or expression.
+        /// </summary>
+        /// <param name="id">Request Id.</param>
+        /// <param name="paramsJson">Request Params Stringified Body.</param>
+        private void HandleFullDocumentSemanticTokens(string id, string paramsJson)
+        {
+            if (!TryParseAndValidateSemanticTokenParams(id, paramsJson, out SemanticTokensParams semanticTokensParams))
+            {
+                return;
+            }
+
+            var uri = new Uri(semanticTokensParams.TextDocument.Uri);
+            var queryParams = HttpUtility.ParseQueryString(uri.Query);
+            var expression = queryParams?.Get("expression") ?? string.Empty;
+
+            if (string.IsNullOrEmpty(expression))
+            {
+                // Empty tokens for the empty expression
+                SendEmptySemanticTokensResponse(id);
+                return;
+            }
+
+            // Monaco-Editor sometimes uses \r\n for the newline character. \n is not always the eol character so allowing clients to pass eol character
+            var eol = queryParams?.Get("eol");
+            eol = !string.IsNullOrEmpty(eol) ? eol : EOL.ToString();
+
+            var tokenTypesToSkip = ParseTokenTypesToSkipParam(queryParams.Get("tokenTypesToSkip"));
```

Comment:
We are planning to combine top down tokenizer (client level) and semantic tokenizer (language server level) to deliver best and efficient user experience when it comes to tokenization and syntax highlighting. We are working on creating a top down tokenizer in monaco-editor that matches TexlLexer. With that, we can recognize many of the token types such as numbers, booleans, string literals, parts of string interpolations, both types of comments at the client level itself. This opens up an opportunity to skip these token types when doing semantic tokenizer thus decreasing the time it takes to compute and process these tokens at the server level and also greatly reduce the size of the response we have to send back to the client. For example, if there is a string that spans 15 lines and given that the encoding implementation creates 15 string tokens for each line to support multi line tokens and also given that each token is encoded to 5 integers according to LSP specification, we can reduce the size of the encoded tokens array by 15 * 5 integers if we could recognize it at the client level and skip it at the language server level. This ability allows us to skip arbitrary token types from the client level. Currently, this is done by passing a list of token types as a value of query param "tokenTypesToSkip" in document URI. Original solution was to use TokenType enum which contains all possible token types as Flags and pass in a bit mask in the document URI and check for the bit status to determine whether to skip token type or not. This would have required changing values in TokenType enum to be powers of 2 but didn't choose do it due to recent regression in this enum that broke old formula bar's syntax highlighting. It is very sensitive. however the list of token types is not huge so this alternative should be fine

Reply:
Ack on benefits of client-side tokenization. But:

1. How will we ensure this new JS implementation stays in sync with C# impl?
2. What happens when they diverge?

Reply:
Folks from VSCode/Monaco recommend combining both Top Down Tokenizer and Semantic Tokenizer for the best tokenization experience. There are various reasons for this but most revolve around efficiency. We have a lexer in powerfx which tokenizes the expression but it doesn't know if an identifier token is a function name or the first name is referring to control or not until the binder. Binder adds semantics to each token. The idea is the same. Client Top down tokenizer acts as a PowerFx Lexer while Language Server adds semantics to each token. Monaco-Editor has two types of tokenization mechanisms (top-down tokenizers and semantic tokenizers). There are certainly many benefits of this combination. However, as you pointed out, there are challenges to this approach which manly revolve around keeping them sync
Both questions touch the same point. 

1) This is definitely a problem but even before these changes, we have been using top down tokenizer  in new formula bar and it has not been correctly set up and also  out of sync with power fx lexer hence there's another work going on to bring up it up to speed with power fx lexer. The cleanest solution for this would be powerfx's implementation in Javascript/Typescript but that is out of scope. Monaco exposes Api to tokenize the expression for all the languages we have registered to it. We are planning to use this to add tests around new top down tokenizer to make sure it is working correctly. The tests would include different type of expressions which would be tokenized using Monaco's API and asserted against expected token types. If the client tokenization goes out of sync then only syntax coloring is affected. The tokenization for any meaningful operations still happen at the server level. Overall usability of the editor is not affected with tokenizer going out of sync. Intellisense, signature help and other features are not affected by client level tokenization. The tests we plan to add would ensure correctness, but it still won't completely solve the problem as any changes in the C# implementation would require us to update the tests as well. How often PowerFx Lexer is updated? The client level tokenizer mostly focuses on simple tokens like numbers, boolean literals, string literals and anything else gets mostly identified as identifiers. How often do we change how basic tokens are recognized? I don't expect that too change that often (probably not all). Note that the list of token types that we want to skip during semantic tokenization would only involve these basic tokens which probably don't change at all.
 
2) This feature to skip certain token types is totally optional. You could always not choose to skip any token type and you will get all token types by default. If client level tokenizer starts to differ and we catch that, we would request language server to send all semantic token types and temporarily disable client level tokenizer until we fix it. That's y i allowed clients to control which tokens show up in the response of semantic tokens request from language server. 

1) & 2) I mentioned in #1 that the cleanest solution would be to implement JS version of powerfx but that it is out of scope. there's another viable better solution which would be less work compared to JS version of powerfx but it would still require large effort in both language server and client code of new formula bar. Language Server Protocol has a concept of exchanging capabilities between client and server when the communication between both is established. We could exchange a wide range of information with the client including the PowerFx lexer rules. The tokenizer in the client is declarative (bunch of regular expressions). This gives us an opportunity to create a declarative version of PowerFx Lexer (TexlLexer) and we can exchange that with the client when the communication is established and client then can setup the top down tokenizer with the declarative rules it got from the server. This basically solves both problems you pointed out above. But this would require a significant effort but something to consider.

Reply:
Resolving this @MikeStall. Let me know if you want to discuss this further. any work to keeping them in sync is out of scope of this pr but this could be solved like i mentioned